### ISSUE TITLE
Add `engine.io` and `ws` engines to [socket] resource 

### ISSUE THREAD

**Issue by Marak** [bigcompany/resources#11]
The `sockets` resource should be able to start different types of socket servers depending on what is needed.

If anyone needs an `engine.io` or `ws` server, they can be added here: https://github.com/bigcompany/resources/tree/master/socket/lib/engines

**Comment by juliankrispel** [comment]
What about SockJs and out of curiousity, how do you imagine the interface for this resource, I guess this isn't finished atm.

**Comment by Marak** [comment]
The `socket` resource is fully functional and tested using socket.io see: https://github.com/bigcompany/resources/blob/master/socket/test/tests.js

Feel free to add `sockjs` or these other engines if you want.

**Comment by juliankrispel** [comment]
Right. First of all, thank you! I'm a big fan of what you're doing here and quite excited by this pattern.

What I haven't completely understood is how you decide which resource methods to expose and which to keep private. My guess is they must be completely reflectable. Would you mind explaining it to me?

**Comment by Marak** [comment]
Cool!

The default behavior for all reflection resources is to expose every method. They should also be able to have resources explicitly passed in.

A small modification to `socket` would allow to explicitly expose resources, see: https://github.com/bigcompany/resources/blob/master/socket/index.js#L33

Moving forward,  once the `auth` branch is merged in we'll implement a permissions system for all resource methods.

**Comment by Marak** [comment]
@juliankrispel - Feel free to join us in our irc room. http://big.vc/irc

**Comment by juliankrispel** [comment]
Thanks! Reading those tests certainly clarified a lot. I'm writing a SockJS adapter now, The socket.io engine doesn't emit events, I have a feeling that's on purpose? ( On a sidenote should I just make a pull request to master or to another branch? )

**Comment by juliankrispel** [comment]
Hi sorry for bothering again, I see now in the tests that you're using callbacks to get data from resources over sockets like this:

``` javascript
  client.emit('creature', 'get', 'korben', function (err, result) {
    t.error(err, 'no error');
    t.type(result, 'object', 'creature is an object');
    t.equal(result.id, 'korben', 'id is korben');
    t.end();
  });
```

I'm not sure how this is at all possible since websockets use JSON for messaging and JSON isn't allowed to contain functions. My point is that I think this resource needs to write as well as read messages. Now there is a chance that I've misunderstood but in case you agree with me, would you be interesting in having the socket resource emit stuff and if so can I help and approximately how would you like it to be implemented. 

PS: I've written a SockJs adapter based on your existing socket.io implementation this morning, will make a pull request as soon as I have some more time to finish it up)

**Comment by Marak** [comment]
Do you have a question? The tests work and are the example syntax to use.

I'm not sure what your confusion is about how socket.io works.

**Comment by juliankrispel** [comment]
What I mean is that in those tests you're embedding functions within socket messages. 

When running those tests they look like they're working callbacks because server and client are running in the same context, but  that's not practical though. In a real situation, client and server won't know about each other so functions embedded in messages are kind of useless, not too mention a security mess. The browser client for socket.io conveniently ignores functions passed into the socket.emit arguments (I think in all clients this should throw an error, I'll file an issue). I've created an [example here](https://github.com/juliankrispel/socketio-example) to illustrate how passed in functions are ignored.

Now in those tests you're using these callbacks to 'get' a creature for example. Again that works in here since we're running both client and server within the same context, but won't work anywhere else and is a misuse of sockets. What I'm saying is that if you want to get stuff from the server the server needs to take the initiative and send a message to the client. I think a more 'reactive' approach would be good. So instead of the client getting a resource like it's http, the server broadcasts the resource whenever it's changed, like in meteor or derby.

**Comment by Marak** [comment]
The example you posted does not prove that functions are ignored and your interpretation of the tests is incorrect.

In your example you are only logging one argument, there are multiple arguments here. I cloned your code, the second argument on this line after `data` is a callback. https://github.com/juliankrispel/socketio-example/blob/master/server.js#L12, and it works as expected.

You can see the callback being used here as well: https://github.com/bigcompany/resources/blob/master/socket/lib/engines/socketio.js#L18

I am unsure what you are trying to ask. If you have a specific question or issue try posting it as code or preferably a failing test.

**Comment by Marak** [comment]
Seem to have missed you in IRC. Jump back in and we can chat.
